### DISCUSSION TITLE
Can multiple stream uploads be active at the same time?

### DISCUSSION THREAD

**Discussion by Niproblema** [jchristn/WatsonTcp#168] [Q&A] [answered]
For example; there is a huge file upload that is going slowly in the first stream. Does that mean that all other streams, which might be significantly faster and more important, have to wait for the first one to finish, or may other streams be processed in parallel?

I am not sure if I am looking at the correct thing, but this excerpt might indicate that due to write lock, only one stream might be active at the time in either direction. Is this correct?
```
        private async Task<bool> SendInternalAsync(WatsonMessage msg, long contentLength, Stream stream, CancellationToken token)
        {
            if (msg == null) throw new ArgumentNullException(nameof(msg));
            if (!Connected) return false;
            
            if (contentLength > 0  && (stream == null || !stream.CanRead))
            {
                throw new ArgumentException("Cannot read from supplied stream.");
            }

            if (token == default(CancellationToken))
            {
                CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(token, _Token);
                token = linkedCts.Token;
            }

            bool disconnectDetected = false;
             
            if (_Client == null || !_Client.Connected)
            {
                disconnectDetected = true;
                return false;
            }
                 
            await _WriteLock.WaitAsync(token).ConfigureAwait(false);

            try
            { 
                await SendHeadersAsync(msg, token).ConfigureAwait(false); 
                await SendDataStreamAsync(contentLength, stream, token).ConfigureAwait(false);

                _Statistics.IncrementSentMessages();
                _Statistics.AddSentBytes(contentLength);
                return true;
            }
            catch (TaskCanceledException)
            {
                return false;
            }
            catch (OperationCanceledException)
            {
                return false;
            }
            catch (Exception e)
            {
                _Settings.Logger?.Invoke(Severity.Error,
                    _Header + "failed to write message to " + _ServerIp + ":" + _ServerPort + ":" +
                    Environment.NewLine +
                    e.ToString() +
                    Environment.NewLine);

                disconnectDetected = true;
                return false;
            }
            finally
            {
                _WriteLock.Release(); 

                if (disconnectDetected)
                {
                    Connected = false;
                    Dispose();
                }
            }
        }
```

**Comment by jchristn** [comment, marked as answer]
Correct, you don't want multiple concurrent writers to a stream. The data would then be interleaved. You would want to use multiple separate connections.

**Comment by Niproblema** [comment]
This makes perfect sense, but it would be super useful for me. For example, I have multiple network streams, users uploading their data, which is proxied to dedicated service. Without concurrent streams a single user might block upload for everyone else indefinitely also.
I don't expect you to fix this, since it's a single client after all, but do you have a suggestion how I could solve that?

**Reply by jchristn** [reply]
Each user is going to have their own connection though, right?  Those connections are all handled in parallel, it's only when you have multiple writers attempting to write to a single connection.

**Reply by Niproblema** [reply]
Yes, each user is their own entity that uploads data through REST. But this data has to be transferred from the gateway service to another dedicated one. At this step only one (Watson)TCP connection is used, purely so I don't have to manage multiple sockets. This would be inconvenient due to firewall and high amount of concurrent uploads. So I intended to use a single TCP connection to stream all the data. What would you do? Use another messaging protocol? Any on your mind?

**Reply by jchristn** [reply]
Ok, I think I got it.  Your clients are connected to some intermediary node (in this case a RESTful endpoint) and this intermediary node has a single TCP connection, using WatsonTcp, to a server.

Yes, in that case, you're going to run into problems for sure; you nailed it.  

What I would recommend is that your intermediary node, as the client, initiates a new TCP connection to the server for each incoming client request.  It's still using the same server port, so you shouldn't have to worry about firewall issues (you're likely only allowing traffic to come *in* on the port on which WatsonTcp is listening, so as long as the destination port matches, you should be fine).

Would that be workable in your situation?

**Reply by Niproblema** [reply]
If server is capable  of receiving one stream from each of up to 1000 clients concurrently, then this should work perfectly. 
Simple solution, I like it. Let's not overengineer.

**Reply by jchristn** [reply]
Awesome.  Are there typically 1000 concurrent clients uploading?  How large are the objects they are sending?  If the TCP setup and CPU overhead of establishing connections becomes a limiting factor, it may make sense to go back to your original idea of having multiple persistent connections between client and server, and load-balancing across those.  Cheers, Joel

**Reply by Niproblema** [reply]
Sometimes. Object size vary from few KB to a few GB. Thanks!!

**Reply by Niproblema** [reply]
One more quick questions. Should large streams be split into multiple chunks manually based on the `.Settings.StreamBufferSize`, or does the library already do that internally?

**Reply by jchristn** [reply]
Hi @Niproblema the ```StreamBufferSize``` only sets the size for the buffer used to read/write from/to underlying streams.  It won't have an impact on the max size supported, though setting it:
1) too small will cause more frequent reads/writes, which could/may cause more latency
2) too large could cause more memory consumption per connection (buffer size is allocated for each connection)

In general I wouldn't expect you would need to sway from the defaults.

**Reply by Niproblema** [reply]
I am having an issue that reading any stream over the `MaxProxiedStreamSize` limit. Anything under works fine.

Line where I issue the stream reading, `args.DataStream.CopyToAsync(Stream, CancellationToken).ConfigureAwait(false);`   immediately causes the client to disconnect. After that a `Cannot access a closed Stream. `exception is thrown.

Would you know what is the correct approach here?

**Reply by Niproblema** [reply]
@jchristn Don't worry I won't bother you anymore, but can you check my last question :)

**Reply by jchristn** [reply]
Hi @Niproblema I will try and spend some time investigating this today.  Cheers

**Reply by Niproblema** [reply]
@jchristn Thanks, I really appreciate it! Let me elaborate, 
I use 
```
        private static MemoryStream GetStream(long mb)
        {
            MemoryStream memoryStream = new MemoryStream();
            memoryStream.Seek(mb * 1024 * 1024, SeekOrigin.Begin);
            memoryStream.WriteByte(0);
            memoryStream.Position = 0;
            return memoryStream;
        }
```
to simulate a large file, then i stream it from client to server or vice versa (fails in both cases). As soon as the recipient tries to read the stream by `args.DataStream.CopyToAsync(Stream, CancellationToken).ConfigureAwait(false);` the client calls disonnect and the copy fails due to stream closing. Let me know if you can replicate.

**Reply by jchristn** [reply]
Having a hard time reproducing this...  Here is a test program I wrote which sets the max proxied stream size to 32.  A client is instantiated which sends 10 messages, where each message is ```i * 10``` bytes in length, so after the third iteration, the underlying data stream is passed.  [Gist](https://gist.github.com/jchristn/8ebd9b8c3449890a07320362e56ebf4f)

And the resultant output:
![image](https://user-images.githubusercontent.com/1421081/145060019-b611956b-186d-4c79-afed-245bd0488fd5.png)

Could you provide an example of this so I can reproduce?

**Reply by Niproblema** [reply]
Hi, sorry for the delay. Eventually I managed to transfer streams over 10GB. Thanks !
